In [1]:
import pandas as pd
import tweepy
import json
import requests
import os
import numpy as np

# Gathering:

### WeRateDogs Twitter Archive:

In [5]:
archive = pd.read_csv('twitter-archive-enhanced.csv'); archive.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

### Tweet Image Predictions:

In [4]:
folder_name = 'tweet_image_predictions'

if not os.path.exists(folder_name):
    os.makedirs(folder_name)
    
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)

In [5]:
with open(os.path.join(folder_name, url.split('/')[-1]), mode = 'wb') as file:
    file.write(response.content)

In [68]:
image_preds = pd.read_csv('./tweet_image_predictions/image-predictions.tsv', sep = '\t'); image_preds.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

### Twitter API:

In [5]:
# initialize the credentials of Twitter Developer Account
consumer_key = '7mBdsNPDYQWk4kgsWMMF3RnQH'
consumer_secret = 'eILB8OwpUKQc7xSb8CB9RfxezhdRHfGdwFcocSYFxsk2IkrbYo'
access_token = '1140993197037895680-VPJik4MsoXr8pTCpXCczdNpfJvV8hB'
access_secret = '37wszdnwEg3rIHt5I55GDNODN5kIFPOf6lGOVa6enQU32'

# authenticate as per tweepy docs
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

# create api object
api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [71]:
# make folder to store the twitter api data
folder_name = 'twitter_api_data'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

# init counter to keep track of tweets collected and of the failed ids
i=1
failed_ids = []

# open file to write json objects from api
with open('./twitter_api_data/tweet_json_no_indent.txt', 'w') as outfile:
    for _ in archive.tweet_id:
        
        # try-except block since few tweet IDs in the archive may have been deleted
        try: 
            tweet = api.get_status(_, tweet_mode = 'extended')
        except:
            failed_ids.append(_)
            i = i+1
            continue
        
        # print the number of tweets collected
        print(f'{_} | {i} of {len(archive.tweet_id)}')
        i = i+1
        
        # dump the json object corresponding to the tweet collected from the api
        json.dump(tweet._json, outfile)
        outfile.write('\n')

892420643555336193 | 1 of 2356
892177421306343426 | 2 of 2356
891815181378084864 | 3 of 2356
891689557279858688 | 4 of 2356
891327558926688256 | 5 of 2356
891087950875897856 | 6 of 2356
890971913173991426 | 7 of 2356
890729181411237888 | 8 of 2356
890609185150312448 | 9 of 2356
890240255349198849 | 10 of 2356
890006608113172480 | 11 of 2356
889880896479866881 | 12 of 2356
889665388333682689 | 13 of 2356
889638837579907072 | 14 of 2356
889531135344209921 | 15 of 2356
889278841981685760 | 16 of 2356
888917238123831296 | 17 of 2356
888804989199671297 | 18 of 2356
888554962724278272 | 19 of 2356
888078434458587136 | 21 of 2356
887705289381826560 | 22 of 2356
887517139158093824 | 23 of 2356
887473957103951883 | 24 of 2356
887343217045368832 | 25 of 2356
887101392804085760 | 26 of 2356
886983233522544640 | 27 of 2356
886736880519319552 | 28 of 2356
886680336477933568 | 29 of 2356
886366144734445568 | 30 of 2356
886267009285017600 | 31 of 2356
886258384151887873 | 32 of 2356
88605416005907251

842163532590374912 | 265 of 2356
842115215311396866 | 266 of 2356
841833993020538882 | 267 of 2356
841680585030541313 | 268 of 2356
841439858740625411 | 269 of 2356
841320156043304961 | 270 of 2356
841314665196081154 | 271 of 2356
841077006473256960 | 272 of 2356
840761248237133825 | 273 of 2356
840728873075638272 | 274 of 2356
840698636975636481 | 275 of 2356
840696689258311684 | 276 of 2356
840632337062862849 | 277 of 2356
840370681858686976 | 278 of 2356
840268004936019968 | 279 of 2356
839990271299457024 | 280 of 2356
839549326359670784 | 281 of 2356
839290600511926273 | 282 of 2356
839239871831150596 | 283 of 2356
838952994649550848 | 284 of 2356
838921590096166913 | 285 of 2356
838916489579200512 | 286 of 2356
838831947270979586 | 287 of 2356
838561493054533637 | 288 of 2356
838476387338051585 | 289 of 2356
838201503651401729 | 290 of 2356
838150277551247360 | 291 of 2356
838085839343206401 | 292 of 2356
838083903487373313 | 293 of 2356
837820167694528512 | 294 of 2356
8374822493

810896069567610880 | 518 of 2356
810657578271330305 | 519 of 2356
810284430598270976 | 520 of 2356
810254108431155201 | 521 of 2356
809920764300447744 | 522 of 2356
809808892968534016 | 523 of 2356
809448704142938112 | 524 of 2356
809220051211603969 | 525 of 2356
809084759137812480 | 526 of 2356
808838249661788160 | 527 of 2356
808733504066486276 | 528 of 2356
808501579447930884 | 529 of 2356
808344865868283904 | 530 of 2356
808134635716833280 | 531 of 2356
808106460588765185 | 532 of 2356
808001312164028416 | 533 of 2356
807621403335917568 | 534 of 2356
807106840509214720 | 535 of 2356
807059379405148160 | 536 of 2356
807010152071229440 | 537 of 2356
806629075125202948 | 538 of 2356
806620845233815552 | 539 of 2356
806576416489959424 | 540 of 2356
806542213899489280 | 541 of 2356
806242860592926720 | 542 of 2356
806219024703037440 | 543 of 2356
805958939288408065 | 544 of 2356
805932879469572096 | 545 of 2356
805826884734976000 | 546 of 2356
805823200554876929 | 547 of 2356
8055206356

777641927919427584 | 768 of 2356
777621514455814149 | 769 of 2356
777189768882946048 | 770 of 2356
776819012571455488 | 771 of 2356
776813020089548800 | 772 of 2356
776477788987613185 | 773 of 2356
776249906839351296 | 774 of 2356
776218204058357768 | 775 of 2356
776201521193218049 | 776 of 2356
776113305656188928 | 777 of 2356
776088319444877312 | 778 of 2356
775898661951791106 | 779 of 2356
775842724423557120 | 780 of 2356
775733305207554048 | 781 of 2356
775729183532220416 | 782 of 2356
775364825476165632 | 783 of 2356
775350846108426240 | 784 of 2356
775085132600442880 | 786 of 2356
774757898236878852 | 787 of 2356
774639387460112384 | 788 of 2356
774314403806253056 | 789 of 2356
773985732834758656 | 790 of 2356
773922284943896577 | 791 of 2356
773704687002451968 | 792 of 2356
773670353721753600 | 793 of 2356
773547596996571136 | 794 of 2356
773336787167145985 | 795 of 2356
773308824254029826 | 796 of 2356
773247561583001600 | 797 of 2356
773191612633579521 | 798 of 2356
7728774959

Rate limit reached. Sleeping for: 115


758740312047005698 | 901 of 2356
758474966123810816 | 902 of 2356
758467244762497024 | 903 of 2356
758405701903519748 | 904 of 2356
758355060040593408 | 905 of 2356
758099635764359168 | 906 of 2356
758041019896193024 | 907 of 2356
757741869644341248 | 908 of 2356
757729163776290825 | 909 of 2356
757725642876129280 | 910 of 2356
757611664640446465 | 911 of 2356
757597904299253760 | 912 of 2356
757596066325864448 | 913 of 2356
757400162377592832 | 914 of 2356
757393109802180609 | 915 of 2356
757354760399941633 | 916 of 2356
756998049151549440 | 917 of 2356
756939218950160384 | 918 of 2356
756651752796094464 | 919 of 2356
756526248105566208 | 920 of 2356
756303284449767430 | 921 of 2356
756288534030475264 | 922 of 2356
756275833623502848 | 923 of 2356
755955933503782912 | 924 of 2356
755206590534418437 | 925 of 2356
755110668769038337 | 926 of 2356
754874841593970688 | 927 of 2356
754856583969079297 | 928 of 2356
754747087846248448 | 929 of 2356
754482103782404096 | 930 of 2356
7544495129

727175381690781696 | 1146 of 2356
727155742655025152 | 1147 of 2356
726935089318363137 | 1148 of 2356
726887082820554753 | 1149 of 2356
726828223124897792 | 1150 of 2356
726224900189511680 | 1151 of 2356
725842289046749185 | 1152 of 2356
725786712245440512 | 1153 of 2356
725729321944506368 | 1154 of 2356
725458796924002305 | 1155 of 2356
724983749226668032 | 1156 of 2356
724771698126512129 | 1157 of 2356
724405726123311104 | 1158 of 2356
724049859469295616 | 1159 of 2356
724046343203856385 | 1160 of 2356
724004602748780546 | 1161 of 2356
723912936180330496 | 1162 of 2356
723688335806480385 | 1163 of 2356
723673163800948736 | 1164 of 2356
723179728551723008 | 1165 of 2356
722974582966214656 | 1166 of 2356
722613351520608256 | 1167 of 2356
721503162398597120 | 1168 of 2356
721001180231503872 | 1169 of 2356
720785406564900865 | 1170 of 2356
720775346191278080 | 1171 of 2356
720415127506415616 | 1172 of 2356
720389942216527872 | 1173 of 2356
720340705894408192 | 1174 of 2356
72005947208178

700518061187723268 | 1387 of 2356
700505138482569216 | 1388 of 2356
700462010979500032 | 1389 of 2356
700167517596164096 | 1390 of 2356
700151421916807169 | 1391 of 2356
700143752053182464 | 1392 of 2356
700062718104104960 | 1393 of 2356
700029284593901568 | 1394 of 2356
700002074055016451 | 1395 of 2356
699801817392291840 | 1396 of 2356
699788877217865730 | 1397 of 2356
699779630832685056 | 1398 of 2356
699775878809702401 | 1399 of 2356
699691744225525762 | 1400 of 2356
699446877801091073 | 1401 of 2356
699434518667751424 | 1402 of 2356
699423671849451520 | 1403 of 2356
699413908797464576 | 1404 of 2356
699370870310113280 | 1405 of 2356
699323444782047232 | 1406 of 2356
699088579889332224 | 1407 of 2356
699079609774645248 | 1408 of 2356
699072405256409088 | 1409 of 2356
699060279947165696 | 1410 of 2356
699036661657767936 | 1411 of 2356
698989035503689728 | 1412 of 2356
698953797952008193 | 1413 of 2356
698907974262222848 | 1414 of 2356
698710712454139905 | 1415 of 2356
69870348362152

684594889858887680 | 1628 of 2356
684588130326986752 | 1629 of 2356
684567543613382656 | 1630 of 2356
684538444857667585 | 1631 of 2356
684481074559381504 | 1632 of 2356
684460069371654144 | 1633 of 2356
684241637099323392 | 1634 of 2356
684225744407494656 | 1635 of 2356
684222868335505415 | 1636 of 2356
684200372118904832 | 1637 of 2356
684195085588783105 | 1638 of 2356
684188786104872960 | 1639 of 2356
684177701129875456 | 1640 of 2356
684147889187209216 | 1641 of 2356
684122891630342144 | 1642 of 2356
684097758874210310 | 1643 of 2356
683857920510050305 | 1644 of 2356
683852578183077888 | 1645 of 2356
683849932751646720 | 1646 of 2356
683834909291606017 | 1647 of 2356
683828599284170753 | 1648 of 2356
683773439333797890 | 1649 of 2356
683742671509258241 | 1650 of 2356
683515932363329536 | 1651 of 2356
683498322573824003 | 1652 of 2356
683481228088049664 | 1653 of 2356
683462770029932544 | 1654 of 2356
683449695444799489 | 1655 of 2356
683391852557561860 | 1656 of 2356
68335797314247

Rate limit reached. Sleeping for: 91


676975532580409345 | 1801 of 2356
676957860086095872 | 1802 of 2356
676949632774234114 | 1803 of 2356
676948236477857792 | 1804 of 2356
676946864479084545 | 1805 of 2356
676942428000112642 | 1806 of 2356
676936541936185344 | 1807 of 2356
676916996760600576 | 1808 of 2356
676897532954456065 | 1809 of 2356
676864501615042560 | 1810 of 2356
676821958043033607 | 1811 of 2356
676819651066732545 | 1812 of 2356
676811746707918848 | 1813 of 2356
676776431406465024 | 1814 of 2356
676617503762681856 | 1815 of 2356
676613908052996102 | 1816 of 2356
676606785097199616 | 1817 of 2356
676603393314578432 | 1818 of 2356
676593408224403456 | 1819 of 2356
676590572941893632 | 1820 of 2356
676588346097852417 | 1821 of 2356
676582956622721024 | 1822 of 2356
676575501977128964 | 1823 of 2356
676533798876651520 | 1824 of 2356
676496375194980353 | 1825 of 2356
676470639084101634 | 1826 of 2356
676440007570247681 | 1827 of 2356
676430933382295552 | 1828 of 2356
676263575653122048 | 1829 of 2356
67623736539290

671542985629241344 | 2042 of 2356
671538301157904385 | 2043 of 2356
671536543010570240 | 2044 of 2356
671533943490011136 | 2045 of 2356
671528761649688577 | 2046 of 2356
671520732782923777 | 2047 of 2356
671518598289059840 | 2048 of 2356
671511350426865664 | 2049 of 2356
671504605491109889 | 2050 of 2356
671497587707535361 | 2051 of 2356
671488513339211776 | 2052 of 2356
671486386088865792 | 2053 of 2356
671485057807351808 | 2054 of 2356
671390180817915904 | 2055 of 2356
671362598324076544 | 2056 of 2356
671357843010908160 | 2057 of 2356
671355857343524864 | 2058 of 2356
671347597085433856 | 2059 of 2356
671186162933985280 | 2060 of 2356
671182547775299584 | 2061 of 2356
671166507850801152 | 2062 of 2356
671163268581498880 | 2063 of 2356
671159727754231808 | 2064 of 2356
671154572044468225 | 2065 of 2356
671151324042559489 | 2066 of 2356
671147085991960577 | 2067 of 2356
671141549288370177 | 2068 of 2356
671138694582165504 | 2069 of 2356
671134062904504320 | 2070 of 2356
67112220491924

667211855547486208 | 2283 of 2356
667200525029539841 | 2284 of 2356
667192066997374976 | 2285 of 2356
667188689915760640 | 2286 of 2356
667182792070062081 | 2287 of 2356
667177989038297088 | 2288 of 2356
667176164155375616 | 2289 of 2356
667174963120574464 | 2290 of 2356
667171260800061440 | 2291 of 2356
667165590075940865 | 2292 of 2356
667160273090932737 | 2293 of 2356
667152164079423490 | 2294 of 2356
667138269671505920 | 2295 of 2356
667119796878725120 | 2296 of 2356
667090893657276420 | 2297 of 2356
667073648344346624 | 2298 of 2356
667070482143944705 | 2299 of 2356
667065535570550784 | 2300 of 2356
667062181243039745 | 2301 of 2356
667044094246576128 | 2302 of 2356
667012601033924608 | 2303 of 2356
666996132027977728 | 2304 of 2356
666983947667116034 | 2305 of 2356
666837028449972224 | 2306 of 2356
666835007768551424 | 2307 of 2356
666826780179869698 | 2308 of 2356
666817836334096384 | 2309 of 2356
666804364988780544 | 2310 of 2356
666786068205871104 | 2311 of 2356
66678179225549

In [14]:
len(failed_ids)

29

In [3]:
#load the json data and store it in a list
data = []
with open('./twitter_api_data/tweet_json_no_indent.txt') as f:    
        for line in f:         
            data.append(json.loads(line))

df_api = pd.DataFrame(data)
#select columns of interest

columns_of_interest = ['id', 'retweet_count', 'favorite_count']
df_api = df_api[columns_of_interest]

## Assessing:

#### Key Points:

- You only want original ratings (no retweets) that have images. Though there are 5000+ tweets in the dataset, not all are dog ratings and some are retweets.

- Assessing and cleaning the entire dataset completely would require a lot of time, and is not necessary to practice and demonstrate your skills in data wrangling. Therefore, the requirements of this project are only to assess and clean at least 8 quality issues and at least 2 tidiness issues in this dataset.

- Cleaning includes merging individual pieces of data according to the rules of tidy data.

- The fact that the rating numerators are greater than the denominators does not need to be cleaned. This unique rating system is a big part of the popularity of WeRateDogs.

- You do not need to gather the tweets beyond August 1st, 2017. You can, but note that you won't be able to gather the image predictions for these tweets since you don't have access to the algorithm used.

> We have the following dataframes:
 `archive`, `image_preds`, `df_api`.

### Assessing `archive`

In [18]:
archive.head(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None  
2                  10   Archie  None    None   None  None

In [27]:
archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [37]:
archive.duplicated().any()

False

In [34]:
archive[archive.retweeted_status_id.notnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
185   856330835276025856                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
...                  ...                    ...                  ...   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
911   757597904299253760                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
949   752309394570878976                    NaN         

In [47]:
archive.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', 'None', 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver',
       'Jim', 'Zeke', 'Ralphus', 'Canela', 'Gerald', 'Jeffrey', 'such',
       'Maya', 'Mingus', 'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey',
       'Lilly', 'Earl', 'Lola', 'Kevin', 'Yogi', 'Noah', 'Bella',
       'Grizzwald', 'Rusty', 'Gus', 'Stanley', 'Alfy', 'Koko', 'Rey',
       'Gary', 'a', 'Elliot', 'Louis', 'Jesse', 'Romeo', 'Bailey',
       'Duddles', 'Jack', 'Emmy', 'Steven', 'Beau', 'Snoopy', 'Shadow',
       'Terrance', 'Aja', 'Penny', 'Dante', 'Nelly', 'Ginger', 'Benedict',
       'Venti', 'Goose', 'Nugget', 'Cash', 'Coco', 'Jed', 'Sebastian',
       'Walter', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Dawn', 'Boomer', 'Cody', 'Rumble', 'Clifford',
       'quite', 'Dewey', 'Scout', 'Gizmo', 'Cooper', 'Harold', 'Shikha',
       'Jamesy', 'Lili', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Nept

In [27]:
archive.source.unique() 

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'],
      dtype=object)

#### Data Quality issues:
- There are 181 retweets as evident from the column `retweet_status_id`.
- `timestamp` column is type `object`.
- `tweet_id` column is type `int64`.
- Some of the tweets with `name` as: **[the, old, actually, just, an, mad, one, not, such, a, None]** are not relating to dogs.
- The `source` column has html unnecessary HTML tags.

#### Data Tidiness issues:
- The columns names `doggo`, `floofer`, `pupper`, `puppo` are not variables. They are possible values.

### Assessing: `image_preds`

In [19]:
image_preds.head(3)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True

- tweet_id is the last part of the tweet URL after "status/" → https://twitter.com/dog_rates/status/889531135344209921
- p1 is the algorithm's #1 prediction for the image in the tweet → Welsh_springer_spaniel
- p1_conf is how confident the algorithm is in its #1 prediction → 46.50%
- p1_dog is whether or not the #1 prediction is a breed of dog → TRUE
- p2 is the algorithm's second most likely prediction → collie
- p2_conf is how confident the algorithm is in its #2 prediction → 15.66%
- p2_dog is whether or not the #2 prediction is a breed of dog → TRUE
- etc.

In [47]:
image_preds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [57]:
image_preds.duplicated().any()

False

In [63]:
image_preds.p1_dog.value_counts()

True     1532
False     543
Name: p1_dog, dtype: int64

#### Data Quality issues:
- `tweet_id` is data type `int64`.
- The above 543 rows have `p1_dog` value as False. They're not dogs.

### Assessing: df_api

In [20]:
df_api.head(3)

id  retweet_count  favorite_count
0  892420643555336193           8141           37445
1  892177421306343426           6031           32180
2  891815181378084864           3993           24265

In [13]:
df_api.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2335 entries, 0 to 2334
Data columns (total 3 columns):
id                2335 non-null int64
retweet_count     2335 non-null int64
favorite_count    2335 non-null int64
dtypes: int64(3)
memory usage: 54.8 KB


In [15]:
df_api.duplicated().any()

False

#### Data Tidiness Issues:
- `df_api` should not be an independent table. It should be joined with `archive`.

# Cleaning:

### Twitter `archive`:

#### Data Quality issues:
- Drop the 181 retweets as evident from the column `retweet_status_id`.
- Convert `timestamp` column from `object` to `datetime`.
- Convert `tweet_id` column from `int64` to `object`.
- Drop the tweets with `name` as **[the, old, actually, just, an, mad, one, not, such, a, None]** since they are not relating to dogs.
- Remove the HTML tags from the entries of the `source` column using `regex`.

#### Data Tidiness issues:
- `melt` the columns `doggo`, `floofer`, `pupper`, `puppo` as they are not variables names but possible values. 

In [6]:
archive.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

In [7]:
archive_clean = archive[archive.retweeted_status_id.isnull()]

In [8]:
archive_clean.timestamp = pd.to_datetime(archive.timestamp)

C:\Users\Dhawal Potdar\Anaconda3\lib\site-packages\pandas\core\generic.py:5096: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


In [9]:
archive_clean.tweet_id = archive_clean.tweet_id.astype(str)

In [10]:
weird_names = ['the', 'old', 'actually', 'just', 'an', 'mad', 'one', 'not', 'such', 'a', 'None', 'Nan']

archive_clean.name = archive_clean.name[~archive_clean.name.isin(weird_names)]

In [11]:
archive_clean.source.unique()

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'],
      dtype=object)

In [12]:
archive_clean.source = archive_clean.source.str.replace(r'(^<.+".)', '')
archive_clean.source = archive_clean.source.str.replace(r'<.+', '')

In [58]:
def replace(x):
    return x.replace('None', '')

archive_clean[['doggo', 'floofer', 'pupper', 'puppo']] = archive_clean[['doggo', 'floofer', 'pupper', 'puppo']].apply(lambda x: replace(x))

archive_clean['dog_stage']= archive_clean['doggo'] + archive_clean['floofer'] + archive_clean['pupper'] + archive_clean['puppo']

archive_clean.dog_stage.value_counts()

C:\Users\Dhawal Potdar\Anaconda3\lib\site-packages\pandas\core\frame.py:3391: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]
C:\Users\Dhawal Potdar\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


                1831
pupper           224
doggo             75
puppo             24
doggopupper       10
floofer            9
doggopuppo         1
doggofloofer       1
Name: dog_stage, dtype: int64

In [59]:
archive_clean.dog_stage = archive_clean.dog_stage.str.replace('doggopupper', 'doggo, pupper')
archive_clean.dog_stage = archive_clean.dog_stage.str.replace('doggopuppo', 'doggo, puppo')
archive_clean.dog_stage = archive_clean.dog_stage.str.replace('doggofloofer', 'doggo, floofer')

archive_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], axis = 'columns', inplace = True)
archive_clean.dog_stage = trial.dog_stage.str.strip().replace('', 'None')

C:\Users\Dhawal Potdar\Anaconda3\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [60]:
archive_clean.dog_stage.value_counts()

None              1831
pupper             224
doggo               75
puppo               24
doggo, pupper       10
floofer              9
doggo, floofer       1
doggo, puppo         1
Name: dog_stage, dtype: int64

In [61]:
archive_clean.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                  timestamp              source  \
0 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name dog_stage  
0                  10   Phineas      None  
1                  10     Tilly      None  
2                  10    Archie      None  
3                  10     Darla      None  
4                  10  Franklin      None

### Cleaning `image_preds`

#### Data Quality issues:
- Convert `tweet_id` to data type `object`.
- Remove the 543 rows that have `p1_dog` value as False. They're not dogs.

In [69]:
image_preds_clean = image_preds.copy()

In [70]:
image_preds_clean.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [71]:
image_preds_clean.tweet_id = image_preds_clean.tweet_id.astype(str);

In [78]:
image_preds_clean = image_preds_clean[image_preds_clean.p1_dog != False]

## Cleaning `df_api`

#### Data Tidiness Issues:
- Join `df_api` with `archive`.

In [84]:
df_api_clean = df_api.copy()

In [85]:
df_api_clean.shape

(2335, 3)

In [89]:
df_api_clean.rename(columns = {'id':'tweet_id'}, inplace = True)

In [91]:
df_api_clean.tweet_id = df_api_clean.tweet_id.astype(str)

### Merging Dataframes:

In [98]:
df = archive_clean.merge(df_api_clean, on = 'tweet_id', how = 'left')

In [99]:
df.columns

Index(['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'dog_stage', 'retweet_count',
       'favorite_count'],
      dtype='object')

In [123]:
tweet_info = df[['tweet_id', 'in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp',
       'source', 'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls']]

dog_info = df[['tweet_id', 'rating_numerator',
       'rating_denominator', 'name', 'dog_stage', 'retweet_count',
       'favorite_count']]

In [103]:
dog_info.head()

tweet_id  rating_numerator  rating_denominator      name  \
0  892420643555336193                13                  10   Phineas   
1  892177421306343426                13                  10     Tilly   
2  891815181378084864                12                  10    Archie   
3  891689557279858688                13                  10     Darla   
4  891327558926688256                12                  10  Franklin   

  dog_stage  retweet_count  favorite_count  
0      None         8141.0         37445.0  
1      None         6031.0         32180.0  
2      None         3993.0         24265.0  
3      None         8296.0         40744.0  
4      None         8995.0         38976.0

In [124]:
dog_info = dog_info.merge(image_preds_clean, on='tweet_id', how = 'left')

In [125]:
dog_info.dropna(inplace = True)

In [126]:
dog_info.reset_index(inplace = True)
dog_info.drop('index', axis = 'columns', inplace = True)

#### Final Dataframes: `dog_info`, `tweet_info`

In [129]:
dog_info.head()

tweet_id  rating_numerator  rating_denominator      name  \
0  892177421306343426                13                  10     Tilly   
1  891815181378084864                12                  10    Archie   
2  891327558926688256                12                  10  Franklin   
3  890971913173991426                13                  10       Jax   
4  890609185150312448                13                  10      Zoey   

  dog_stage  retweet_count  favorite_count  \
0      None         6031.0         32180.0   
1      None         3993.0         24265.0   
2      None         8995.0         38976.0   
3      None         1982.0         11439.0   
4      None         4099.0         26949.0   

                                           jpg_url  img_num             p1  \
0  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg      1.0      Chihuahua   
1  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg      1.0      Chihuahua   
2  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg      2.0         basset   
3  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg      1.0    Appenzeller   
4  https://pbs.twimg.com/media/DFwUU__XcAEpyXI.jpg      1.0  Irish_terrier   

    p1_conf p1_dog                p2   p2_conf p2_dog  \
0  0.323581   True          Pekinese  0.090647   True   
1  0.716012   True          malamute  0.078253   True   
2  0.555712   True  English_springer  0.225770   True   
3  0.341703   True     Border_collie  0.199287   True   
4  0.487574   True      Irish_setter  0.193054   True   

                            p3   p3_conf p3_dog  
0                     papillon  0.068957   True  
1                       kelpie  0.031379   True  
2  German_short-haired_pointer  0.175219   True  
3                    ice_lolly  0.193548  False  
4     Chesapeake_Bay_retriever  0.118184   True

In [130]:
tweet_info.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                  timestamp              source  \
0 2017-08-01 16:23:56+00:00  Twitter for iPhone   
1 2017-08-01 00:17:27+00:00  Twitter for iPhone   
2 2017-07-31 00:18:03+00:00  Twitter for iPhone   
3 2017-07-30 15:58:51+00:00  Twitter for iPhone   
4 2017-07-29 16:00:24+00:00  Twitter for iPhone   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  
0  https://twitter.com/dog_rates/status/892420643...  
1  https://twitter.com/dog_rates/status/892177421...  
2  https://twitter.com/dog_rates/status/891815181...  
3  https://twitter.com/dog_rates/status/891689557...  
4  https://twitter.com/dog_rates/status/891327558...